contributors/design-proposals/api-machinery/customresource-conversion-webhook.md

```diff
@@ -0,0 +1,859 @@
+
+* As a cluster owner, when I upgrade to the version of Kubernetes that supports CRD multiple versions, but I don't use the new feature, my existing CRDs work fine.  I can roll back to the previous version without any special action.
+
+
+## Summary of Changes
+
+1. A CRD object now represents a group/kind with one or more versions.
+
+2. The CRD API (CustomResourceDefinitionSpec) is extended as follows:
+
+    1. It has a place to register 1 webhook.
+
+    2. it holds multiple "versions".
+
+    3. Some fields which were part of the .spec are now per-version; namely Schema, Subresources, and AdditionalPrinterColumns.
+
+3. A Webhook Host is used to do conversion for a CRD.
+
+    4. CRD authors will need to write a Webhook Host that accepts any version and returns any version.
+
+    5. Toolkits like kube-builder and operator-sdk are expected to provide flows to assist users to generate Webhook Hosts.
+
+
+## Detailed Design
+
+
+### CRD API Changes
+
+The CustomResourceDefinitionSpec is extended to have a new section where webhooks are defined: 
+
+```golang
+// CustomResourceDefinitionSpec describes how a user wants their resource to appear
+type CustomResourceDefinitionSpec struct {
+  Group string
+  Version string
+  Names CustomResourceDefinitionNames
+  Scope ResourceScope
+  // This optional and correspond to the first version in the versions list
+  Validation *CustomResourceValidation
+  // Optional, correspond to the first version in the versions list 
+  Subresources *CustomResourceSubresources
+  Versions []CustomResourceDefinitionVersion
+  // Optional, and correspond to the first version in the versions list  
+  AdditionalPrinterColumns []CustomResourceColumnDefinition
+
+  Conversion *CustomResourceConversion
+}
+
+type CustomResourceDefinitionVersion struct {
+  Name string
+  Served Boolean
+  Storage Boolean
+  // These three fields should not be set for first item in Versions list
+  Schema *JSONSchemaProp
+  Subresources *CustomResourceSubresources
+  AdditionalPrinterColumns []CustomResourceColumnDefinition
+}
+
+Type CustomResourceConversion struct {
+  // Conversion strategy, either "nop” or "webhook”. If webhook is set, Webhook field is required.
+  Strategy string
+
+  // Additional information for external conversion if strategy is set to external
+  // +optional
+  Webhook *CustomResourceConversionWebhook
+}
+
+type CustomResourceConversionWebhook {
+  // ClientConfig defines how to communicate with the webhook. This is the same config used for validating/mutating webhooks.
+  ClientConfig WebhookClientConfig
+}
+```
+
+### Defaulting
+
+In case that there is no versions list, a single version with values defaulted to top level version will be created. That means a single version with a name set to spec.version.
+All newly added per version fields (schema, additionalPrinterColumns or subresources) will be defaulted to the coresponding top level field except for the first version in the list that will remain empty.
```

Comment:
Because of backward compatibility explained down there. We cannot duplicate data and we cannot remove top level fields so we do not use first version's fields. This will be cleaned-up in v1.